### PR TITLE
feat(cuda): add fused IF/LIF-Linear kernels

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary

- add experimental fused CUDA/CuPy IF-Linear and LIF-Linear operators for single- and multi-step inputs
- avoid materializing the intermediate spike tensor and rematerialize neuron state for first-order backward
- add numerical, gradient, stream, fake-tensor, and torch.compile coverage
- add a reproducible benchmark against unfused Torch, CuPy, and Triton paths

## Validation

- Ruff and compileall passed
- RTX 3090: 8 CUDA tests passed
- LIF, T=1, M=512, K=N=4096, 1% spike rate: fused 0.367 ms vs CuPy dense 1.083 ms; fused_256 was the stable winner
- ponytail triage removed 27 redundant lines before submission

## Summary by CodeRabbit

- **New Features**
  - Added experimental CUDA-accelerated IF-Linear and LIF-Linear operations.
  - Supports single-step and multi-step inputs, optional bias, configurable reset and decay behavior, surrogate gradients, transposed weights, and CUDA thread settings.
  - Supports automatic differentiation, compiled models, and meta-device shape propagation.
  - Returns output sequences and final membrane states.

- **Tests**
  - Added coverage for outputs, gradients, reset modes, CUDA streams, and compilation compatibility.

- **Benchmarks**
  - Added GPU performance and peak-memory comparisons across supported implementations and configurations.

### CODE REVIEW

File: test/activation_based/test_cuda_neuron_linear.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            sg=surrogate.Sigmoid(),
+        )
+    torch.cuda.current_stream().wait_stream(stream)
+    torch.testing.assert_close(y, y_ref, rtol=1e-4, atol=2e-5)
+    torch.testing.assert_close(v_next, v_ref)
+    grad_y, grad_v = torch.randn_like(y), torch.randn_like(v_next)
+    torch.autograd.backward((y, v_next), (grad_y, grad_v))
+    torch.autograd.backward((y_ref, v_ref), (grad_y, grad_v))
+    for actual, expected in zip((x, v, weight), refs):
+        torch.testing.assert_close(actual.grad, expected.grad, rtol=2e-4, atol=2e-5)
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply test/activation_based/test_cuda_neuron_linear.py]: Fixed in b53cedc6: every gradient-comparison zip call now uses strict=True.

File: benchmark/binary_kernel/benchmark_neuron_linear.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+                            with torch.no_grad():
+                                expected = methods["torch_dense"]()
+                                for name, fn in methods.items():
+                                    actual = fn()
+                                    torch.testing.assert_close(
+                                        actual[0], expected[0], rtol=1e-4, atol=1e-4
+                                    )
+                                    torch.testing.assert_close(
+                                        actual[1], expected[1], rtol=1e-5, atol=1e-6
+                                    )
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/binary_kernel/benchmark_neuron_linear.py:244]: Not applicable: none of the torch, CuPy, Triton, IF-Linear, or LIF-Linear methods mutates the supplied v. Each returns a distinct next state. The benchmark validation completes successfully on CUDA.

File: spikingjelly/activation_based/cuda_kernel/neuron_linear.py [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+        v_out.data_ptr(),
+        np.int32(T),
+        np.int32(M),
+        np.int32(K),
+        np.int32(N),
+        np.float32(1.0 / tau),
+        np.float32(v_threshold),
+        np.float32(v_reset),
+        np.int32(bias is not None),
+    )
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply spikingjelly/activation_based/cuda_kernel/neuron_linear.py:281]: Not changing: both paths use the same FP32 reciprocal. The fused launch computes np.float32(1.0 / tau), while functional.lif_step_cupy and functional.lif_multi_step_cupy also compute decay as 1.0 / tau before launching CUDA. The tau 3.1 threshold-boundary regression passes on CUDA.

File: benchmark/binary_kernel/benchmark_neuron_linear.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+                                for name, fn in methods.items():
+                                    timing, peak_bytes = _measure(
+                                        fn,
+                                        tensors,
+                                        (grad_y, grad_v),
+                                        training,
+                                        args.warmup,
+                                        args.iters,
+                                        args.rounds,
+                                    )
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/binary_kernel/benchmark_neuron_linear.py:264]: Not applicable: measured methods do not mutate captured v. Neuron functions and fused custom ops return v_out separately. Resetting v would add work without changing the starting state.

File: test/activation_based/test_cuda_neuron_linear.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    v = torch.randn(3, 16, device="cuda")
+    weight_t = torch.randn(16, 8, device="cuda")
+
+    def f(x, v, weight_t):
+        return lif_linear(x, v, weight_t, threads=128)
+
+    expected = f(x, v, weight_t)
+    actual = torch.compile(f, fullgraph=True)(x, v, weight_t)
+    torch.testing.assert_close(actual[0], expected[0])
+    torch.testing.assert_close(actual[1], expected[1])
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply test/activation_based/test_cuda_neuron_linear.py:223]: Not changing: torch.compile with fullgraph=True raises on a graph break rather than silently falling back. Forward compile coverage stays focused here; eager backward correctness is covered for IF and LIF across single-step, multi-step, hard-reset, and soft-reset cases.

File: benchmark/binary_kernel/benchmark_neuron_linear.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/binary_kernel/benchmark_neuron_linear.py:244]: Not changing: grad mode only controls whether autograd context is registered. The custom-op forward kernel and outputs are identical. Dedicated CUDA tests cover backward numerical equivalence, and training benchmark execution exercises backward.

File: benchmark/binary_kernel/benchmark_neuron_linear.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    for _ in range(20):
+        mid = (low + high) / 2
+        spikes, _ = _neuron_torch(base + mid, v, surrogate.heaviside, neuron)
+        if spikes.float().mean().item() < density:
+            low = mid
+        else:
+            high = mid
+    x = base + (low + high) / 2
+    spikes, _ = _neuron_torch(x, v, surrogate.heaviside, neuron)
+    return x, spikes.float().mean().item()
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/binary_kernel/benchmark_neuron_linear.py:71]: Not changing: each JSON case records target_spike_rate and measured actual_spike_rate, so saturated searches are not silently mislabeled. The current bounds cover the measured workload scope.

File: benchmark/binary_kernel/benchmark_neuron_linear.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    baseline_mad = statistics.median(abs(x - baseline_median) for x in baseline_rounds)
+    winners = []
+    for name, result in measurements.items():
+        if not name.startswith("fused_"):
+            continue
+        rounds = result["timing"]["rounds_ms"]
+        wins = sum(x < y for x, y in zip(rounds, baseline_rounds))
+        gap = baseline_median - result["timing"]["median_ms"]
+        if wins >= 4 and gap > max(baseline_mad, result["timing"]["mad_ms"]):
+            winners.append(name)
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply benchmark/binary_kernel/benchmark_neuron_linear.py:174]: Not changing: wins >= 4 is intentionally strict. Four rounds require 4 out of 4 wins; the default five rounds require 4 out of 5. This is achievable and matches the stable-winner criterion.

File: test/activation_based/test_cuda_neuron_linear.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    v = torch.randn(5, 64, device="cuda")
+    weight = torch.randn(33, 64, device="cuda")
+    stream = torch.cuda.Stream()
+    with torch.cuda.stream(stream):
+        y, v_next = lif_linear(x, v, weight.t().contiguous(), threads=256)
+        y_ref, v_ref = _reference(
+            x,
+            v,
+            weight,
+            None,
```
[the remark this answers is not in the document]
Comment (AllenYolk) [reply test/activation_based/test_cuda_neuron_linear.py:164]: Already addressed: test_lif_linear_without_bias_and_nondefault_stream backpropagates through fused and reference outputs and compares x, v, and weight gradients with bias=None.
